src/wallet/wallet.cpp

```diff
@@ -86,6 +86,12 @@ bool CWallet::ActivateSaplingWallet(bool memOnly)
 {
     if (m_sspk_man->SetupGeneration(m_spk_man->GetHDChain().GetID(), true, memOnly)) {
         LogPrintf("%s : sapling spkm setup completed\n", __func__);
+        // Just to be triple sure, if the version isn't updated, set it.
+        if (!CanSupportFeature(WalletFeature::FEATURE_SAPLING)) {
+            if (!SetMinVersion(WalletFeature::FEATURE_SAPLING)) {
+                LogPrintf("%s : ERROR: wallet cannot upgrade to sapling features. Try to upgrade using the 'upgradewallet' RPC command\n", __func__);
```

Comment:
This requires `pwalletMain->cs_wallet` locked.
There's some unit test (e.g. `wallet_zkeys_tests`) getting here without holding the mutex (`WriteCryptedSaplingZkeyDirectToDb` via `SetupSPKM`).

Reply:
fixed